chore(categories): add category API description

joggienl opened the pull request:
Add API documentation for the Category object.
Some posible discussion points are:

- How are the nested categories being returned?
 If there is nesting multiple levels deep, it might not be wise to return all the nested categories inline. My suggestion is to return a list (array) with all the nested categories. I'm not sure how this affects the progress!
- How are the nested speech challenges being returned?
 Same as with the nested categories but since this is only one level deep I would suggest to just include them in the result.
- How are the (eventually) nested categories and or speech challenges being returned after updating or creating an category? Can we leave out fields (like progress) when creating a category?
 My suggestion would be to show as minimal as possible on create. This feature will not be used in a fronted directly (possible in a sort of CMS solution).

joggienl committed:
chore(categories): add categorie API description

Lehren commented on docs/api/categories.md:
```diff
@@ -0,0 +1,265 @@
+# Cagegories
```
Categories*

Lehren commented on docs/api/categories.md:
```diff
@@ -0,0 +1,265 @@
+
+### Response
+
+The resonse is a JSON list with categories.
```
response

Lehren committed:
docs: fix typos

kvdb commented on docs/api/categories.md:
```diff
@@ -0,0 +1,265 @@
+    "color": "#00f",
+    "imageUrl": "https://api.itslanguage.nl/download/UKbsMpBsXaJUsBbK",
+    "iconUrl": "https://api.itslanguage.nl/download/GdExSbs-ZVNnQUUe",
+    "categories": ["category_1_1", "category_1_2"],
```
subcategories would be more suitable, however, I would not like to include this field. I rather query /categories again with extra filter: parent_category=category_1. Why? Because adding a new category in the database would require updating the parent category as well as adding the new category. Or extra queries need to be done in python at the time a user GETs /categories. Alternatively, like cgen has the API: we could provide the link where to GET the subcategories (HATEOAS). Then again, I like consistency and we don't provide links to other API calls in the output, except for download links.

dcsiebe replied:
Updating the parent when creating a subcategory is not necessarily a bad thing. If this saves extra queries for subcategories it is preferred. 
Providing HATEOAS links would also require to query subcategories, no?

kvdb replied:
Not necessarily. It could be something like https://api.itslanguage.nl/categories?parent_category=categorie_1. That requires no extra queries. However, the developer could also think of that himself...

dcsiebe replied:
Ah, yes, that doesn't require any additional queries, I thought you wanted to link them with an absolute url (including the subcategory id), I guess I just misunderstood.
As far as the urls go, I think I'd rather use url paths for this instead of a query param; I like to keep query params as filters and apply hackable urls. Getting subcategories would imo look something like this: `/categories/categorie_1/categories`.

dcsiebe replied:
I guess the bottom line is; do we need this information directly (on the same page the categories are listed)? If so, it is good to include this info, if not, an additional query is best. @joggienl could you elaborate on this?

joggienl replied:
I've removed the "categories" field in favour of the "parent" field.
Now, my suggestion to get all subcategories belonging to parent_x:

```http
GET /categories/:category/subcategories HTTP/1.1
```

Using "subcategories" in stead of "categories" on the end tells more about what you're fetching (imho).

HATEOAS is a nice concept. We might want to consider that for future releases!

dcsiebe replied:
If categories are to be nested at multiple levels, I still think `/categories/:category/categories/:category/categories` would be nicer. But it's not worth discussing over.

joggienl replied:
That could be `/categories/:category/subcategories/:category/subcategories` I guess?

kvdb commented on docs/api/categories.md:
```diff
@@ -0,0 +1,265 @@
+    "imageUrl": "https://api.itslanguage.nl/download/UKbsMpBsXaJUsBbK",
+    "iconUrl": "https://api.itslanguage.nl/download/GdExSbs-ZVNnQUUe",
+    "categories": ["category_1_1", "category_1_2"],
+    "speechChallenges": [],
```
Same reason as above. Rather not.

kvdb commented on docs/api/categories.md:
```diff
@@ -0,0 +1,265 @@
+    "iconUrl": "https://api.itslanguage.nl/download/GdExSbs-ZVNnQUUe",
+    "categories": ["category_1_1", "category_1_2"],
+    "speechChallenges": [],
+    "progress": {
```
Now we're mixing personal progress with the generic category data in one API call. Not sure if that's a good idea. It might be, but it's a new idea so I have to think about it. Because it makes caching API calls harder, I lean towards an extra API call to get your personal progress, but I'm not fully leaning one way or the other yet.

dcsiebe replied:
If the progress is to be shown on the same page as the overview of categories it is good to include it. It might affect caching, but it would save a round trip and an extra api call implementation.

joggienl replied:
Had some offline discussion about progress. Best is to get it out for now and re-discuss this when we're ready to implement it.

kvdb commented on docs/api/categories.md:
```diff
@@ -0,0 +1,265 @@
+  "updated": "2017-01-12T09:38:20Z",
+  "name": "Category 1.1",
+  "description": "Super duper.",
+  "color": "",
```
Rather use JS value null in this case (and also the cases below)

kvdb commented:
nested categories, I'm not sure. IMHO api calls should be simple and snappy. No problem to request a couple of them to fill the screen. Although, if we know clients won't be having 100's of categories, it might make sense to just return 'm all at once. Not my preference however.

dcsiebe commented on docs/api/categories.md:
```diff
@@ -0,0 +1,265 @@
+`color`            | `string`  | **Optional** A color, preferably in RGB format.
+`image`            | `blob`    | **Optional** An image to show with the category.
+`icon`             | `blob`    | **Optional** An icon to show with the category.
+`categories`       | `object`  | **Optional** An array with category identifiers for nesting categories in a category.
```
I think a subcategory can only be created when a parent category already exists. I don't see why a subcategory can be set in a category when creating the category. So in short, I think this param should be removed.

kvdb replied:
+1

dcsiebe commented on docs/api/categories.md:
```diff
@@ -0,0 +1,265 @@
+`color`            | `string`  | **Optional** A color, preferably in RGB format.
+`image`            | `blob`    | **Optional** An image to show with the category.
+`icon`             | `blob`    | **Optional** An icon to show with the category.
+`categories`       | `object`  | **Optional** An array with category identifiers for nesting categories in a category.
```
Same as the other comment. I think the back-end should manage this automatically based on the creation/deletion of subcategories.

dcsiebe reviewed (changes requested):
Github won't allow me to do a review without this comment. I might be doing it wrong.

joggienl committed:
refactor(categories): remove 'categories' and 'progress', add parent

Changing the responsibility for nesting. If a category changes, you only
have to update the child.

Progress is removed in favor of development speed. This concept needs to
be thought over and will be added in a later release.

joggienl commented:
I think we also need to add a call that get's all the (speech) challenges belonging to an category. This in favour of including the speech challenges in the result of requesting (multiple) categories.
Could be like this:

```http
GET /categories/:category/speech HTTP/1.1
```

dcsiebe reviewed (approved):
Looks fine to me, let's try it like this.

joggienl committed:
refactor: describe how to get child categories, fix some typos

Describe how to get all categories linked to a single parent. Also fix
some typo's and remove the nested SpeechChallenges (only the identifiers
will be present in the result).
